### PR TITLE
[FSDP2] Apply reduce_dtype to the non-DP-mesh gradient all-reduce

### LINKED ISSUE #190626: [FSDP2] MixedPrecisionPolicy reduce_dtype is not applied to the TP-mesh Partial -> Replicate gradient all-reduce
## Bug description

Under FSDP2 + tensor parallelism (2D `dp x tp` mesh), `MixedPrecisionPolicy(param_dtype=torch.bfloat16, reduce_dtype=torch.float32)` applies `reduce_dtype` only to FSDP's own reduce-scatter. The `Partial() -> Replicate()` all-reduce over the TP mesh, which FSDP itself issues during post-backward for every TP-replicated parameter (e.g. norm weights under sequence parallelism, replicated biases, MoE router weights), runs in the gradient's own dtype, i.e. `param_dtype` bf16. The user asks for an fp32 reduction, FSDP accepts the policy silently, and half of the 2D reduction still happens in bf16.

**Impact: every torchtitan run with `tensor_parallel_degree > 1` on shipped defaults hits this, including the published Llama 3.1 70B and 405B demonstration configs (both tp=8): their norm-weight gradients are TP-reduced in bf16 while the config promises, and only permits, fp32.**

The `MixedPrecisionPolicy` docstring says `reduce_dtype` "specifies the dtype for gradient reduction (i.e. reduce-scatter or all-reduce)" (`torch/distributed/fsdp/_fully_shard/_fsdp_api.py`). The TP all-reduce is an all-reduce that FSDP issues, and it is not covered.

## Minimal repro (4 GPUs, dp=2 x tp=2; also reproduces at 2 GPUs with mesh (1, 2))

```python
# torchrun --standalone --nproc_per_node=4 repro.py
import math
import torch
import torch.nn as nn
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.fsdp import fully_shard, MixedPrecisionPolicy
from torch.distributed.tensor import DTensor, Replicate, Shard, distribute_tensor

mesh = init_device_mesh("cuda", (2, 2), mesh_dim_names=("dp", "tp"))
tp = mesh["tp"]
tp_rank = tp.get_local_rank()

class Scale(nn.Module):                       # norm-style TP-Replicate param
    def __init__(self, d):
        super().__init__()
        self.w = nn.Parameter(torch.ones(d))
    def forward(self, x):
        return x * self.w

m = Scale(1024).cuda()
m.w = nn.Parameter(distribute_tensor(m.w.detach(), tp, [Replicate()]))
fully_shard(m, mesh=mesh["dp"], mp_policy=MixedPrecisionPolicy(
    param_dtype=torch.bfloat16, reduce_dtype=torch.float32))

x_local = torch.zeros(8, 1024, device="cuda", dtype=torch.bfloat16)
x_local[0, :] = 1.0 if tp_rank == 0 else 2.0 ** -10   # both exact in bf16
m(DTensor.from_local(x_local, tp, [Shard(0)], run_check=False)).to_local().sum().backward()

g = m.w.grad.full_tensor().double().cpu()[0]
mant = 2.0 * math.frexp(float(g))[0]          # dp SUM/AVG is a power of two
assert mant == 1.0 + 2.0 ** -10, f"TP all-reduce ran in bf16: mantissa {mant}"
```

The per-rank TP partial gradients are exactly `1.0` and `2^-10`, both exact in bf16. Their sum `1 + 2^-10` is exact in fp32 and rounds (no tie) to exactly `1.0` in bf16, whose spacing at 1.0 is `2^-8`. The mantissa of the final (fp32, master-weight) gradient is therefore a bit-exact witness of the collective's dtype. Observed: mantissa `1.0` (the entire low term is lost). Expected with `reduce_dtype=float32`: `1.0009765625`. Deterministic across seeds; identical on torch 2.9.1+cu128 and a 2.12 nightly (2026-04-08) on 4x H100; the mechanism is unchanged on current main.

## Root cause

- `_fsdp_param.py`, `unsharded_grad_data` -> `_get_grad_inner_tensor`: for a DTensor parameter, the gradient is redistributed to the parameter's placements on non-DP mesh dims (`grad.redistribute(...)`). Partial -> Replicate on the TP dim is an all-reduce executed in `grad.dtype` = `param_dtype`. No cast happens anywhere on this path.
- `_fsdp_param_group.py`, `post_backward`: collects `unsharded_grad_data`, i.e. the TP all-reduce has already happened before anything `reduce_dtype`-aware runs.
- `_fsdp_collectives.py`, `foreach_reduce`: `reduce_dtype = reduce_dtype or grad_dtype` and the reduce-scatter buffer allocation are the first and only places `reduce_dtype` takes effect.
- Decisive internal inconsistency: on the gradient-accumulation path, `to_accumulated_grad_if_needed` upcasts the unsharded gradient to `reduce_dtype` before `unsharded_accumulated_grad_data` calls the same `_get_grad_inner_tensor`. So the exact same model reduces the TP dim in fp32 when `set_requires_gradient_sync(False)` was used and in bf16 on the plain path. Two orderings of the same two operations, one of them wrong.

## Reachability

Verified against the shipped torchtitan configs of the era the flagship demonstrations were published (commit db6b39ac, Oct 2024) as well as current main; every trigger condition is a default:

- `train_configs/llama3_70b.toml` and `llama3_405b.toml` both ship `tensor_parallel_degree = 8` with `data_parallel_shard_degree = -1`. (`llama3_8b.toml` ships tp=1 and is unaffected; the trigger is purely tp > 1, model size is irrelevant.)
- `config_manager.py` of that era already locked `--training.mixed_precision_reduce` to `default="float32", choices=["float32"]`; users could not opt out of the fp32-reduce promise then and cannot now.
- `parallelize_llama.py` of that era placed `attention_norm` and `ffn_norm` under `SequenceParallel()`, which replicates the weight over the TP mesh and leaves its gradient Partial, pending the all-reduce that FSDP issues.
- torch v2.5.0 (`_composable/fsdp/_fsdp_param.py:674-682`) carries the same unguarded `Partial -> Replicate` redistribute as current main, so the mechanism spans that era through today. Runtime confirmation is on 2.9.1 and a 2026-04 nightly; the v2.5 evidence is source-level.

The dp size does not matter: the TP all-reduce runs in `_get_grad_inner_tensor` even with a size-1 DP mesh, so single-node tp>1 runs are equally affected.

This is a shipped-default configuration in torchtitan: `mixed_precision_param="bfloat16"` and `mixed_precision_reduce="float32"` (the only permitted value) feed `MixedPrecisionPolicy` directly, `enable_sequence_parallel=True` by default, and the norm weight is placed Replicate on the TP mesh with its pending backward all-reduce explicitly delegated to FSDP (`decoder_sharding.py`: "R if SP (pending BWD AR handled by FSDP)"). Any torchtitan run with `tensor_parallel_degree > 1` and `data_parallel_shard_degree > 1` on defaults reduces its `attention_norm.weight` and `ffn_norm.weight` gradients over the TP group in bf16 while every other parameter is reduced in fp32.

Scope: only TP-Replicate parameters are affected (norm weights, replicated biases, routers); Shard-placed weights have no Partial to reduce. The error is one bf16 rounding of a TP-size-way sum, about 2^-9 relative, not a sign or magnitude corruption. It is exactly the precision the user paid `reduce_dtype=float32` to keep, on the parameters where low-precision reduction is known to matter most, and nothing warns: a user auditing the reduce-scatter buffer dtype sees fp32 and is misled.

## Versions

Verified failing on torch 2.9.1+cu128 and 2.12.0.dev20260408+cu128 (4x H100, NCCL). Mechanism present in current main.

## Related issues (searched, distinct)

- #125191 introduced the accumulation-path upcast (the half that is correct); the plain path never got the same treatment.
- #186998 (reduce_dtype resolving to None after a frozen first forward) and #143297 (clamp reduce_dtype in lazy init) touch dtype resolution, not the DTensor redistribute.

*This report was authored with assistance from Claude; the mechanism, repro, and exact-value oracle were human-reviewed.*

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy

### PR DESCRIPTION
**Pull request by bogatyy**
Fixes #190626

## Summary

`MixedPrecisionPolicy(reduce_dtype=...)` is documented to apply to gradient reduction (reduce-scatter or all-reduce), but it only took effect at the reduce-scatter buffer allocation in `foreach_reduce`. The `Partial -> Replicate` all-reduce that `_get_grad_inner_tensor` issues over non-DP mesh dims (e.g. for a TP-replicated norm weight under sequence parallelism) ran in the gradient's own dtype, so with `param_dtype=bf16` and `reduce_dtype=fp32` half of the 2D reduction still happened in bf16. See the issue for the bit-exact witness repro and the torchtitan shipped-default reachability.

The fix upcasts the gradient to `reduce_dtype` at the top of `_get_grad_inner_tensor`, mirroring what `to_accumulated_grad_if_needed` already does on the gradient-accumulation path (introduced in #125191). The accumulation path was already correct; the plain path now takes the same two operations in the same order.

## Why this placement

- It covers all three gradient sources uniformly (`unsharded_grad_data`, `unsharded_accumulated_grad_data` where it is a no-op since the grad is already fp32, and `unsharded_zero_grad_data`).
- It keeps the collected gradient dtypes uniform when fresh and accumulated gradients meet in one post-backward. `foreach_reduce` asserts a single gradient dtype and `fsdp.chunk_cat` requires uniform input dtypes, so a selective per-parameter upcast would trip both; today a mixed post-backward (some parameters accumulated in fp32, others fresh in bf16) would already hit that assert, and this change removes that latent failure as well.
- Memory cost is transient and matches what the accumulation path already pays: the bf16 gradient reference is dropped immediately after collection in `post_backward` and the fp32 copy is freed when `unsharded_grads.clear()` runs after the reduce-scatter copy-in.

## Test

`test_reduce_dtype_tp_grad_all_reduce` (2+ GPUs, `dp x tp` mesh with `tp=2`): a Replicate-on-TP scale parameter receives TP-local partial gradients of exactly `1.0` and `2^-10`, both exact in bf16. Their sum is exact in fp32 but rounds back to `1.0` in bf16 (spacing `2^-8` at 1.0), so the final master-weight gradient being exactly `1.0 + 2^-10` is a bit-exact witness that the collective ran in fp32. The test fails on main and passes with this change.

```
pytest test/distributed/_composable/fsdp/test_fully_shard_mixed_precision.py -k test_reduce_dtype -v
```

*This PR was authored with assistance from Claude; the fix design and test oracle were human-reviewed.*

### CODE REVIEW

File: torch/distributed/fsdp/_fully_shard/_fsdp_param.py [hunk 1]
```diff
@@ -1149,6 +1149,15 @@ def unsharded_zero_grad_data(self) -> torch.Tensor:
         return self._get_grad_inner_tensor(torch.zeros_like(self.unsharded_param))
 
     def _get_grad_inner_tensor(self, grad: torch.Tensor) -> torch.Tensor:
+        if self.reduce_dtype is not None and grad.dtype != self.reduce_dtype:
```
Comment (RohitRathore1) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param.py:1154]: when `reduce_dtype` is set and the input grad is an `AsyncCollectiveTensor, grad.to(self.reduce_dtype)` implicitly resolves the ACT before reaching the explicit `grad.wait()` guard below, making that branch unreachable for this case. The guard still works when `reduce_dtype` is `None`.

Comment (bogatyy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param.py:1154]: Good catch. Hoisted the ACT resolution above the upcast in 6d54efb so the wait stays explicit and the guard stays reachable; this also covers the spmd branch, which previously wrapped the grad without checking for an ACT.

**Comment by bogatyy** [comment]
@weifengpy ready for review: 208 checks green; the remaining ciflow/torchtitan and ciflow/inductor workflows are awaiting maintainer approval (bottom of the PR page). The torchtitan run seems worth having here since the shipped tp>1 configs are the default-path trigger for #190626.

**Review by weifengpy — COMMENTED** [review]
thanks for drafting the PR carefully, and most importantly, pointing out the issue itself is extreme valuable

I want to focus on TP reduce dtype discussion first: https://github.com/pytorch/pytorch/issues/190626#issuecomment-5042381822
